xonsh/main.py

```diff
@@ -252,10 +253,58 @@ def premain(argv=None):
     return args
 
 
+def _failback_to_other_shells(argv):
+    args = None
+    try:
+        args = premain(argv)
+    except Exception:
+        pass
+    # only failback for interactive shell; if we cannot tell, treat it
+    # as an interactive one for safe.
+    if hasattr(args, 'mode') and args.mode != XonshMode.interactive:
+        sys.exit(1)
```

Comment:
Ah, I didn't know `sys.exit(1)` is not suitable here .. and what the concern about it? I use sys.exit so that xonsh can exit with code 1. I have checked `return 1` cannot made the exit code to 1.

I'm ok with a just return without updating exit code - tho that will make `xonsh -c 'list(?'` exited with exit code 0.